### ISSUE TITLE
Support skipping auto-registration for particular files via configuration code

### ISSUE THREAD

**Issue by timriley** [dry-rb/dry-system#46]
In https://github.com/dry-rb/dry-system/pull/45#issuecomment-296590276, @solnic, you mentioned this:

> We may also consider adding a configuration option ie config.auto_registration_exclude { |dir| dir.include?('entities') } or something like that.

I'm looking to start adding this now, so it can complement the magic comments behaviour and then we can consider a new dry-system release.

However, I feel hesitant to add another setting like `config.auto_registration_exclude`, for a few reasons:

1. Firstly, dry-configurable won't make it easy to have a block as the value for the setting like your example. Since it uses blocks for nested settings, we'd have to ask users to provide a proc as the value, which doesn't feel as natural
2. I'm also wary about "configuration-creep". If we can avoid a setting at the top-level of the container, I think that'd be good.
3. The user may not want this setting to apply to all their auto-registration paths.
4. Lastly, and perhaps most critically, the auto-registrar class is a setting _itself_, and if the default auto-registrar was replaced, it may not even want to pay any attention to a setting like `auto_registration_exclude`.

So instead of adding another container setting, I wonder if we could reuse the `.auto_register!(path, &block)` method we alreay have, just extending the block so it supports auto-registration configuration for that particular path. So instead of:

```ruby
auto_register!('lib') do |component|
  # some custom initialization logic
  constant = component.loader.constant
  constant.create
end
```

We make it this:

```ruby
auto_register!('lib') do |config|
  # Both of these configuration blocks are optional

  config.instance do |component|
    # some custom initialization logic
    constant = component.loader.constant
    constant.create
  end

  config.exclude do |component|
    component.path =~ /entities/
  end
end
```

What do you think? Since this configuration is coupled to our default `Dry::System::AutoRegistrar`, it feels like a more appropriate place for it.

**Comment by solnic** [comment]
@timriley oh this is a really nice idea!

**Comment by timriley** [comment]
Resolved by #47
